### PR TITLE
Fix `@compiled/babel-plugin` to not silently skip `cssMap()` extraction in a few common edge-cases.

### PR DESCRIPTION
**Pull request by kylorhall-atlassian**
### What is this change?

Fixes https://github.com/atlassian-labs/compiled/issues/1642

Resolves two issues revolving around `cssMap` not extracting which could lead to production issues as Runtime and Extraction are handled differently with little to no testability in this area.
1. When `cssMap` is called after it's used, eg. `<div css={styles.root}>; const styles  = cssMap({ root: … });`, styles are never extracted
2. When the root `cssMap` object is passed to css, eg. `<div css={styles}>; const styles  = cssMap(…);`, styles are never extracted and this results in no typescript error — built via https://github.com/atlassian-labs/compiled/pull/1774 and merged into this branch already.

### Why are we making this change?

Both of these seemingly acceptable code examples passed full CI, TS, Linting, Babel, etc., but styles were simply never displayed to the user.

```tsx
import { cssMap } from "@compiled/react";
export default () => <div css={styles.root} />;
const styles = cssMap({ root: { padding: 8 } });
```

```tsx
import { cssMap } from "@compiled/react";
const styles = cssMap({ root: { padding: 8 } });
export default () => <div css={styles} />;
```

These should both worth now and be tested, with the downside that this _could_ result in breaking changes by "fixing" these currently erroring paths…

---

### PR checklist

I have...

- [x] Updated or added applicable tests
- [x] ~Updated the documentation in `website/`~ n/a
- [x] Added a changeset (if making any changes that affect Compiled's behaviour)

### CODE REVIEW

File: packages/babel-plugin/src/babel-plugin.ts [hunk 1]
```diff
@@ -178,6 +244,11 @@ export default declare<State>((api) => {
               );
             }
           }
+
+          // Build our `state.compiledImports[…]` state now that we've traversed JSX pragmas
+          path.traverse<State>(buildCompiledImportsCache, this);
+          // Build our `state.cssMap[…]` before we traverse the JSXElements which consume it
+          path.traverse<State>(buildCssMapState, this);
```
Comment (kylorhall-atlassian) [inline packages/babel-plugin/src/babel-plugin.ts:251]: I don't know if this is the best way to guarantee the order of things—just stuffing them into `Program.enter`, but it appears to work.

This could definitely be improved and traversion could be reduced by 1-5x if desired…

File: packages/babel-plugin/src/css-map/__tests__/index.test.ts [hunk 2]
```diff
@@ -36,6 +36,25 @@ describe('css map basic functionality', () => {
     ]);
   });
 
+  it('should transform css map even when the styles are defined below the component', () => {
```
Comment (kylorhall-atlassian) [inline packages/babel-plugin/src/css-map/__tests__/index.test.ts:39]: Before, this test would actually throw a recursion error from finding member expressions—which maybe my recent change introduced the recursion error actually, so it helped me debug this.

**Comment by kylorhall-atlassian** [comment]
Hitting an edge-case in the built tests, may circle back.

File: packages/babel-plugin/src/babel-plugin.ts (hunk 1 above)
Comment (kylorhall-atlassian) [reply packages/babel-plugin/src/babel-plugin.ts:251]: This is wrong…in tests and other places, these aren't guaranteed to be at the Program scope, so this breaks our own tests.

File: packages/babel-plugin/src/utils/css-builders.ts [hunk 3]
```diff
@@ -688,11 +691,47 @@ function extractMemberExpression(
...
+
+  if (bindingIdentifier) {
+    /**
+     * If we don't yet have a `meta.state.cssMap[…]` built yet, build it and cache it, eg. in this scenario:
+     * ```tsx
+     * const Component = () => <div css={styles.root} />
+     * const styles = cssMap({ root: { padding: 0 } });
+     * ```
+     *
+     * If we don't hit this, we put it into `ignoreMemberExpressions` to ignore on future runs.
```
Comment (kylorhall-atlassian) [inline packages/babel-plugin/src/utils/css-builders.ts:703]: Not the most performant thing to do, but I couldn't find a better way that works with `cssMap` defined in different scopes, eg. not hoisted within Jest files as `packages/react/src/create-strict-api/__tests__/index.test.tsx` failed if I injected this into `Program` => `path.traverse()` instead.

File: packages/react/src/create-strict-api/__tests__/generics.test.tsx
Comment (kylorhall-atlassian) [inline packages/react/src/create-strict-api/__tests__/generics.test.tsx:1]: Ideally, this wouldn't be required, but the way these tests are built seems like a false-positive where we allow `it()` level scope for `css` and `cssMap`, yet in the real-world, you would never have `cssMap()` that wasn't at the Program scope as it throws errors.

What's happening is `state.cssMap['styles']` is being called and sometimes this is a `css` and other time it's a `cssMap`…this forces them to be unique at least.

**Review by liamqma — APPROVED** [review]
